### PR TITLE
Added Copy Clean Link item to renderer view context menu

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/25845

- Added menu item to renderer context menu to clean url
- Added strict cleaning for urls that goes through:
  - debouncer (potentially debouncing many levels)
  - query filter
  - (the normal clean URL sanitization)
- Little refactoring to move tests files from `brave/test/BUILD.gn` to `/brave/browser/net`

image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Use https://fmarier.github.io/brave-testing/copy-clean-link.html to confirm that all test cases are working as expected.

### CODE REVIEW

**Review by goodov — APPROVED** [review]

**Review by fmarier — CHANGES_REQUESTED** [review]
It's a little more complicated than that (and that's the reason it wasn't in the MVP). This is [mentioned briefly in this section of the spec](https://docs.google.com/document/d/1ea7eF3s0WTsTlmMPr8LYp5KUyNg8eK4omjizo27qvJE/edit#heading=h.utvf3nnnswm0), but briefly, the URL also needs to go through:
- the debouncer
- the query filter
- HTTPS upgrades

All of these already happen on navigations so we get them for free when we copy the current tab URL, but if we're copying a link, we need to apply all of these URL modifications first.

**Comment by spylogsster** [comment]
> It's a little more complicated than that (and that's the reason it wasn't in the MVP). This is [mentioned briefly in this section of the spec](https://docs.google.com/document/d/1ea7eF3s0WTsTlmMPr8LYp5KUyNg8eK4omjizo27qvJE/edit#heading=h.utvf3nnnswm0), but briefly, the URL also needs to go through:
> 
> * the debouncer
> * the query filter
> * deAMP
> * HTTPS upgrades
> 
> All of these already happen on navigations so we get them for free when we copy the current tab URL, but if we're copying a link, we need to apply all of these URL modifications first.

done

**Review by SergeyZhukovsky — APPROVED** [review]
android tests ++

File: browser/ui/browser_commands.cc [hunk 1]
```diff
@@ -185,15 +188,38 @@ void CloseWalletBubble(Browser* browser) {
...
 
   ui::ScopedClipboardWriter scw(ui::ClipboardBuffer::kCopyPaste);
   scw.WriteText(base::UTF8ToUTF16(sanitized_url.spec()));
 }
 
+void CopyLinkWithStrictCleaning(Browser* browser, const GURL& url) {
+  GURL final_url;
+  // Apply debounce rules.
+
+  auto* debouce_service =
```
Comment (fmarier) [inline browser/ui/browser_commands.cc:204]: typo: `debounce_service`

**Review by fmarier — CHANGES_REQUESTED** [review]
Just a small typo to fix. Otherwise it looks good.

Can you confirm that it works with these two test cases? https://fmarier.github.io/brave-testing/copy-clean-link.html

File: browser/net/brave_query_filter.cc [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  // then join the remaining query parameters, untouched, back into
+  // a single query string.
+  const std::vector<std::string> input_kv_strings =
+      SplitString(query, "&", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+  std::vector<std::string> output_kv_strings;
+  int disallowed_count = 0;
+  for (const std::string& kv_string : input_kv_strings) {
+    const std::vector<std::string> pieces = SplitString(
+        kv_string, "=", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+    const std::string& key = pieces.empty() ? "" : pieces[0];
```
Comment (goodov) [inline browser/net/brave_query_filter.cc:83]: I think we can pass `StringPiece` as inputs, use `SplitStringPiece` version everywhere, do all calculation and return `optional<std::string>` only if changes are required.

File: browser/net/brave_query_filter.cc [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    return base::JoinString(output_kv_strings, "&");
+  } else {
+    return query;
+  }
+}
+
+}  // namespace
+
+absl::optional<GURL> ApplyQueryFilter(const GURL& original_url) {
+  const std::string query = original_url.query();
```
Comment (goodov) [inline browser/net/brave_query_filter.cc:104]: `query_piece()`

File: browser/net/brave_query_filter.cc [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  const std::string spec = original_url.spec();
```
Comment (goodov) [inline browser/net/brave_query_filter.cc:105]: `const std::string&`

File: browser/ui/browser_commands.cc (hunk 1 above)
Comment (spylogsster) [reply browser/ui/browser_commands.cc]: done

**Comment by spylogsster** [comment]
> Just a small typo to fix. Otherwise it looks good.
> 
> Can you confirm that it works with these two test cases? https://fmarier.github.io/brave-testing/copy-clean-link.html

yes, it works

File: browser/net/brave_query_filter.cc (hunk 4 above)
Comment (spylogsster) [reply browser/net/brave_query_filter.cc]: done

File: browser/net/brave_query_filter.cc (hunk 2 above)
Comment (spylogsster) [reply browser/net/brave_query_filter.cc]: done, thanks

File: browser/net/brave_query_filter.cc (hunk 3 above)
Comment (spylogsster) [reply browser/net/brave_query_filter.cc]: done

**Review by goodov — APPROVED** [review]

**Review by fmarier — APPROVED** [review]

File: browser/net/brave_query_filter.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_NET_BRAVE_QUERY_FILTER_H_
+#define BRAVE_BROWSER_NET_BRAVE_QUERY_FILTER_H_
+
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "url/gurl.h"
```
Comment (iefremov) [inline browser/net/brave_query_filter.h:10]: can be a forward declaration

File: browser/ui/browser_commands.h [hunk 6]
```diff
@@ -22,7 +23,8 @@ void ShowBraveVPNBubble(Browser* browser);
 void ToggleBraveVPNButton(Browser* browser);
 void OpenBraveVPNUrls(Browser* browser, int command_id);
 void OpenIpfsFilesWebUI(Browser* browser);
-void CopyCleanLink(Browser* browser);
+void CopyCleanLink(Browser* browser, const GURL& url);
+void CopyLinkWithStrictCleaning(Browser* browser, const GURL& url);
```
Comment (iefremov) [inline browser/ui/browser_commands.h:27]: would be great to have comments for these functions, b/c it is not really clear what is clean and what is strictly clean

File: browser/ui/browser_commands.cc [hunk 7]
```diff
@@ -185,15 +188,37 @@ void CloseWalletBubble(Browser* browser) {
...
-                           ->SanitizeURL(browser->tab_strip_model()
-                                             ->GetActiveWebContents()
-                                             ->GetVisibleURL());
+                           ->SanitizeURL(url);
 
   ui::ScopedClipboardWriter scw(ui::ClipboardBuffer::kCopyPaste);
   scw.WriteText(base::UTF8ToUTF16(sanitized_url.spec()));
 }
 
+void CopyLinkWithStrictCleaning(Browser* browser, const GURL& url) {
```
Comment (iefremov) [inline browser/ui/browser_commands.cc:200]: it feels like we need to apply cleaning only to `http(s)://` links? Probably somewhere should be a quick return for non-interesting links

File: browser/net/brave_query_filter_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/net/brave_query_filter.h"
+
+#include "testing/gmock/include/gmock/gmock.h"
```
Comment (iefremov) [inline browser/net/brave_query_filter_unittest.cc:8]: not used? and `url/origin.h`?

**Review by iefremov — APPROVED** [review]

File: browser/net/brave_query_filter.h (hunk 5 above)
Comment (spylogsster) [reply browser/net/brave_query_filter.h]: done, thanks

File: browser/ui/browser_commands.cc (hunk 7 above)
Comment (spylogsster) [reply browser/ui/browser_commands.cc:204]: There are no restrictions by http(s). if we want to add it then I think this should be explained in UI as well, @fmarier wdyt?

File: browser/net/brave_query_filter_unittest.cc (hunk 8 above)
Comment (spylogsster) [reply browser/net/brave_query_filter_unittest.cc]: removed

**Review by bsclifton — APPROVED** [review]
LGTM

File: browser/ui/browser_commands.cc (hunk 7 above)
Comment (fmarier) [reply browser/ui/browser_commands.cc:204]: Good catch, we should limit this to http/https URLs. @spylogsster would you mind filing a follow-up issue for this?

**Pull request merged**